### ISSUE TITLE
Do not type check .js files in referenced node_modules under `--checkJs` and `--maxNodeModuleJsDepth` > 0

### ISSUE THREAD

**Issue by mhegazy** [Microsoft/TypeScript#23219]
- Create a `jsconfig.json`
- Set `checkJs` to `true`
- run `tsc --p ./jsconfig.json`

Expected:
no errors in `node_modules\*\*.js`. 

There is no really good fix for these issues. i would suggest not checking these files

**Comment by apalaniuk** [comment]
I'm guessing this isn't prioritized? I'd love to be able to use `checkJs` as a type lint step vs. simply IDE errors for some JS projects that can't be converted to TypeScript for organizational reasons (for now). I suppose a workaround might be to use `@checkJs` directives in source files, which seems a bit cumbersome/prone to accidental exclusion.  

Not sure if this is open to outside contributions, but if so, it seems like this could respect the effective include/exclude rules rather than be against `node_modules` explicitly.

**Comment by sandersn** [comment]
@apalaniuk The best contribution you could is to add to this issue. The OP is pretty detail free. I'd like to see:

1. A simple, self-contained repro, with full actual and expected behaviour.
2. Multiple use cases with actual/expected behaviour.
3. A discussion of backward compatibility considerations -- what does this mean for people who already have a jsconfig?

If you can provide any of these, it will help move the issue along; if we can decide *what* to do, I think doing it will be pretty easy.

**Comment by Raynos** [comment]
I've run into this issue and thought I could fix it with adding exclude to my jsconfig.json

```
+  "exclude": ["node_modules"]
```

I was suprised to find that `tsc -p jsconfig.json` still checkes `node_modules` even though `exclude` is explicitely set in the jsconfig

**Comment by Raynos** [comment]
A reproduction can be found on this commit for this git repo ( https://github.com/Raynos/fake-s3/commit/282fa39134611857759c1b6a81ccc5cf5d75ccd3 ).

Running `npm run tsc` outputs 400 typescript errors for `node_modules/*` but zero errors for my source code ( index.js with jsdoc ).

**Comment by adomoshe** [comment]
Any update on this?

**Comment by Raynos** [comment]
My workaround is `tsc -p jsconfig.json --maxNodeModuleJsDepth 0`.

For some reason it ignore `maxNodeModuleJsDepth` in jsconfig.json but setting it 0 explicitely fixes it for me.

**Comment by adomoshe** [comment]
Thanks, do you know how to fix this through react-scripts start and build or tsconfig.json? I fixed my problem locally but my real problem is when I upload to the server

**Comment by Raynos** [comment]
i don't compile my javascript. it's javascript, I use `tsc` to "lint" my jsdoc.

**Comment by jespertheend** [comment]
> My workaround is `tsc -p jsconfig.json --maxNodeModuleJsDepth 0`.

This does nothing for me. I'm guessing it's because I have directly imported a bunch of node modules with `import "../node_modules/some-module/dist.js"`

**Comment by thelinuxlich** [comment]
The `--maxNodeModuleJsDepth 0` trick is what solves this issue but I think it deserves more attention.

**Comment by lieser** [comment]
I had a similar issue. I use the TypeScript compiler for type checking my JavaScript Project, but wanted errors in external code to be ignored, without having to modify the external files (e.g. by adding a `@ts-nocheck` at the beginning).

Thanks to https://dev.to/15five/how-to-temporarily-ignore-errors-during-a-typescript-migration-doe I came across `tsc-silent` (https://github.com/evolution-gaming/tsc-silent), which is a wrapper around the TypeScript compiler with a supress option.
Which in my brief test so far works for me.

Maybe that workaround is helpful for others too who find this issue.

**Comment by RyanCavanaugh** [comment]
I don't think we have a realistic path forward here. The jsconfig-specific behaviors are going to be quite breaky to change, and we don't really recommend jsconfig anymore regardless -- it's just better to set these projects up via `tsconfig.json`.

**Comment by jespertheend** [comment]
What about a feature that allows you to configure files which are globally `@ts-ignore`d? I frequently run into a situation where I want to use some third party library with tons of files. And my jsconfig/tsconfig being more strict than whatever that library uses means that I'll end up with tons of unactionable errors/warnings. Enabling `checkJs` is one such example, but there's other options like `exactOptionalPropertyTypes` that have a similar effect.

My usual go-to solution is to use a script I wrote which simply adds a `// @ts-ignore` comment at the top of all .js/.ts files in a specific directory. This works fine but it feels a bit odd that there's no other workaround. I would have expected this to be a standard typescript feature.

**Comment by RyanCavanaugh** [comment]
I think that's reasonable in principle. What we see in real repos is that the definition of what's "my" code and what's not gets blurry really quickly.

**Comment by jespertheend** [comment]
Then it sounds like the solution would be to add highly granular `compilerOptions` controls on a per file basis 😅
